class_name Weapon
extends Node2D

signal energy_used(amount)

@export var fire_rate := 0.3
@export var damage := 10
@export var type := ""

var is_right = false

var _firing = false
var _on_cooldown = false
var _cooldown_timer: Timer


func setup(timer: Timer):
	_cooldown_timer = timer


func _process(_delta):
	var mouse = get_global_mouse_position()
	look_at(mouse)

	_set_weapon_direction(mouse)
	_on_cooldown = _cooldown_timer.time_left > 0.0

	if _firing and not _on_cooldown:
		show()

		_fire()

		_cooldown_timer.start(fire_rate)

	elif not _firing and not _on_cooldown:
		hide()


func fire(pressed: bool):
	_firing = pressed


func _set_weapon_direction(pos: Vector2):
	is_right = pos.x > global_position.x

	_update_weapon_direction()


func _update_weapon_direction():
	pass


func _fire():
	pass
